tools/clutils/isn/saga-isn.1
.\" Man page for SAGA Information System Navigator CLI
.TH SAGA-ISN "1" "January 2010" "SAGA-ISN " "User Commands"
.SH NAME
saga-isn \- navigates the information system via a SAGA API
.SH SYNOPSIS
.B saga-isn \fR
[options] <model> <entity type> ([filter] <entity type>) [attributes]
.SH DESCRIPTION
.\" Add any additional description here
.PP
Navigate from a starting entity set (defined by all members of some entity type that match a filter) via a chain of related entity types to 
a final one. At each hop a filter may be used to restrict the size of the set. 
.\" The special keywords "up" and "down" may be used to navigate a self relationship. 
A filter must be enclosed in square brackets to distinguish it from an entity type. For most 
shells filter arguments will need to be quoted. Finally a list of attributes may be given to display only the values of those attributes. 
Otherwise all attributes are displayed for each member of the set at the end of the chain. The attribute list must be enclosed in 
parentheses and attributes should be separated by commas and/or spaces.  

.SH PARAMETERS
.TP
<model>
The name of the information model to use
.TP
<entity type>
The name of the type of an entity to start from or navigate to
.TP
<filter>
A filter to restrict the number of entities in the set. 
The filter string uses SQL92 
syntax as if it were part of a WHERE clause acting to select from a single table. It must be enclosed in square brackets. Multi-valued attributes are 
treated as a set of values.

Only the following operators are permitted in expressions not involving multi-valued attributes: IN, 
LIKE, AND, OR, NOT, =, >=, >, <=, <, <> in addition to column names, parentheses, column values 
as single quoted strings, numeric values and the comma. For a multi-valued attribute, the name of the 
attribute MUST have the keyword ALL or ANY immediately before it, unless comparison with a set 
literal is intended. For each part of the expression, the attribute name MUST precede the literal value.
The LIKE operator matches string patterns:

'%xyz' matches all entries with trailing xyz

'xyz%' matches all entries with leading xyz

'%xyz%' matches all entries with xyz being a substring

The ESCAPE keyword can be used with LIKE in the normal way.
Column names are not case sensitive but values are.

The operators >=, >, <= and < are not supported when applied to strings.
Column names in the the filter string are matched against the attribute names of the entity.
If values are specified as numeric values and not in single quotes, the values will be converted from 
string to numeric for comparison.

Attributes may be multi-valued. If a filter string does not have the correct syntax to accept multi-
valued attributes, and an entity has more than one value for an attribute mentioned in the filter, that  
entity MUST be rejected.

.TP
<attributes>
This may be used to restrict the set of displayed attributes. It is a space and/or comma separated list enclosed in parentheses

.SH OPTIONS
.TP
\fB\-t, --trace\fR
display count after each hop and list the related entity sets
.TP
\fB\-\-url\fR \fIURL\fR
the URL of the information system to use. This overrides any environment variables.
.TP
\fB\-h, --help\fR
display help and exit
.TP
\fB\-n\fR \fINUM\fB \-\-count\fR \fINUM\fR
the NUM of services to display. The default is unlimited.
.SH Examples
.TP
saga-isn  -n 10 -t glue1 Service "[Type = 'org.glite.RTEPublisher']" Site Cluster ComputingElement  "(FreeCPUs)"
.TP
saga-isn  -n 5 -t glue1 Service "[Type = 'org.glite.RTEPublisher']" Site "(Name, Latitude)"


CHANGES
                            SAGA C++ Core Libraries  
                    - The Simple API for Grid Applications -
                          http://www.saga-project.org


  Changes with SAGA Core Version 1.7                                    T.B.D
  ---------------------------------------------------------------------------
  o 

  Changes with SAGA C++ Version 1.6                         1. September 2011
  ---------------------------------------------------------------------------
  o Updated logging system
  o Implemented packaging mechanisms 
  o Complete list of tickets: http://bit.ly/nBMtep


  Changes with SAGA C++ Version 1.5.4                          19. April 2011
  ---------------------------------------------------------------------------
  o Fixes compatibility issue with boost::filesystem V3
    see: http://bit.ly/dFfrYO
  o http://faust.cct.lsu.edu/trac/saga/ticket/706
  o Some other minor fixes
  
    
  Changes with SAGA C++ Version 1.5.3                        10. January 2011
  ---------------------------------------------------------------------------
  o Lots of improvements in the documentation of the programming 
    examples and the programming example code itself
  o Addressed some Boost compatibility issues
  o Fixed tickets: http://bit.ly/h9XhIw


  Changes with SAGA C++ Version 1.5.2                      28. September 2010
  ---------------------------------------------------------------------------
  o Added support for Boost >= 1.43 to the build-system
  o Disabled the engine monitoring thread until the race cond. has been found
  o Fixed compilation problem with  external boost::asio
  o Fixed compilation problem with  external boost::filesystem and boost <= 1.35
  o Fixed problems that were cause by a bug in boost::filesystem 1.44
  o file.read() fixed in the default file adaptor
  o Lost of minor improvements and fixes


  Changes with SAGA C++ Version 1.5.1                         05. August 2010
  ---------------------------------------------------------------------------
  o Bugfix Release - fixes build system issues (include path precedence)
  o A detailed list of all closed tickets for this release can be 
    found here: http://tinyurl.com/2g2e8lb


  Changes with SAGA C++ Version 1.5                             09. July 2010
  ---------------------------------------------------------------------------
  o A detailed list of all closed tickets for this release can be 
    found here: http://tinyurl.com/2fa6euv
    

  Changes with SAGA C++ Version 1.4.2                          16. April 2010
  ---------------------------------------------------------------------------
  o We have completely restructured our code repositories. We have split up 
    SAGA into a "Core" (Engine, Packages, Default Adaptors) and into the 
    middleware adaptors. SAGA Core still lives in the same code repository,
    but the adaptors have been moved to:
    
    https://svn.cct.lsu.edu/repos/saga-adaptors.
    
    Beginning with this release, we will release Core, Python Bindings and 
    Adaptors seperately. This allows us to have individual release cycles
    for individual adaptors and components. One of the major proeblems we 
    had in the past was, that e.g. a single bug in the SSH adaptor lead to
    a completely new saga release. This won't happen anymore: now you will
    be able to download just the new release of that SSH adaptor and add it
    to your existing SAGA installation. 
    
    What do all these changes and restructuring mean for you? For you it means
    that you can't install SAGA from one source archive anymore. Yes, it 
    might have been convenient but it also caused a lot of confusions. If you
    want to install saga, you have to download and install "SAGA Core" first.
    After that, you can pick the adaptors you want and maybe the Python 
    bindings, download and install them. All Core, Adaptor and Bindings 
    releases can be found on the SAGA C++ download page:
    
    http://saga.cct.lsu.edu/software/cpp/download
    
  o Besides restructuring our code, we also found enough time to fix some
    of the bugs that were reported. Here's a list of recently closed tickets:
    
    http://tinyurl.com/y63aole
  

  Changes with SAGA C++ Version 1.4.1                       17. February 2010
  ---------------------------------------------------------------------------
   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/ycozhef

  Changes with SAGA C++ Version 1.4                         01. December 2009
  ---------------------------------------------------------------------------
   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/yk3f8yc

  Changes with SAGA C++ Version 1.3.3                        01. October 2009
  ---------------------------------------------------------------------------
   o EMERGENCY BUGFIX RELEASE
     Fixes broken context handling 
     Some small improvements in the ssh job adaptor

  Changes with SAGA C++ Version 1.3.2                      28. September 2009
  ---------------------------------------------------------------------------
   o We have improved our project structures. In order to create higher-
     quality releases, we agreed upon a set of strict release policies and
     a new, more conservative release branch. More information can be 
     found here: http://faust.cct.lsu.edu/trac/saga/wiki/Development
     This is the FIRST release coming out of this new branch! 

   o A list of fixed bugs and added features can be found here:
     http://tinyurl.com/ye2qusr

  Changes with SAGA C++ Version 1.3.1                         24. August 2009
  ---------------------------------------------------------------------------
  o A list of fixed bugs and added features can be found here:
    http://tinyurl.com/y8p4jng

  Changes with SAGA C++ Version 1.3                             22. June 2009

tools/clutils/isn/saga-isn.cpp
//  Distributed under the Boost Software License, Version 1.0. (See accompanying 
//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)

#include <iostream>
#include <vector>
#include <stdlib.h>
#include <libgen.h>

#include "saga/saga/packages/isn/entity_data_set.hpp"
#include "saga/saga/packages/isn/entity_data.hpp"
#include "saga/saga.hpp"

/**
 * Takes the program arguments and gives back a pointer to the next one for each call to next()
 */
class Args {

    private:

        std::vector<std::string> m_argv;
        int m_pos;

    public:

        Args(int argc, char* argv[]) :
            m_pos(0) {
            for (int i = 1; i < argc; i++) {
                m_argv.push_back(argv[i]);
            }
        }

        std::string * next() {
            if (m_pos < m_argv.size()) {
                return &m_argv[m_pos++];
            } else {
                return 0;
            }
        }

};

/**
 * Split the string str into the provided vector. The delimiter set defaults to a space.
 */
void tokenize(const std::string& str, std::vector<std::string>& tokens, const std::string& delimiters = " ") {
    std::string::size_type lastPos = str.find_first_not_of(delimiters, 0); /* skip leading delimiters */
    std::string::size_type pos = str.find_first_of(delimiters, lastPos); /* then look for next delimiter */

    while (std::string::npos != pos || std::string::npos != lastPos) {
        tokens.push_back(str.substr(lastPos, pos - lastPos)); /* store token */
        lastPos = str.find_first_not_of(delimiters, pos);
        pos = str.find_first_of(delimiters, lastPos);
    }
}

void printTrace(const saga::isn::entity_data_set& eds, const std::string entityType, const std::string filter) {
    std::cout << entityType << " [" << filter << "] has " << eds.get_entity_count() << " -> ";
    std::vector<std::string> res = eds.list_related_entity_names();
    std::cout << res[0];
    for (int i = 1; i < res.size() - 1; i++) {
        std::cout << ", " << res[i];
    }
    if (res.size() > 1) {
        std::cout << " or " << res[res.size() - 1];
    }
    std::cout << std::endl << std::flush;
}

/**
 * Function to return value from a string. A check is made that there is nothing else after valid input apart from white space.
 */
template<class T>
bool from_string(T& t, const std::string& s) {
    std::istringstream iss(s);
    if ((iss >> t).fail()) {
        return false;
    }
    int eof;
    iss >> eof;
    return iss.fail() && iss.eof();
}


/**
 * Main program
 */
int main(int argc, char *argv[]) {
#if !defined(SAGA_HAVE_PACKAGE_ISN)
  std::cerr << "saga-isn: saga has been compiled without the isn "
               "package, bailing out" << std::endl;
  return -1;
#else
    std::string usage("Usage: " + std::string(basename(argv[0]))
            + " [options] <model> <entity type> ([filter] <entity type>) [attributes]");

    try {
        Args args(argc, argv);

        bool trace = false;
        int maxCount = std::numeric_limits<int>::max();
        std::string serverUrl = "";
        std::string* nextArgP = args.next();
        while (nextArgP && nextArgP->at(0) == '-') {
            if (*nextArgP == "-h" || *nextArgP == "--help") {
                std::cout << std::endl;
                std::cout << usage << std::endl;
                std::cout << "Options:" << std::endl;
                std::cout << "   --trace, -t          display count after each hop and related entity sets"
                        << std::endl;
                std::cout << "   --count, -n <value>  the number of entities to display - default unlimited"
                        << std::endl;
                std::cout << "   --url <server url>   this overrides any environment variables" << std::endl;
                std::cout << "   --help, -h           print this message" << std::endl;
                std::cout << "Parameters:" << std::endl;
                std::cout << "    <model>             name of the model - must be known to the adaptor" << std::endl;
                std::cout << "    <entity type>       name of an entity type" << std::endl;
                std::cout << "    <filter>            filter enclosed in square brackets" << std::endl;
                std::cout
                        << "    <attributes>        attributes to display separated by spaces and or commas and enclosed in parantheses - default all"
                        << std::endl;
                std::cout << std::endl;
                return 0;
            } else if (*nextArgP == "-t" || *nextArgP == "--trace") {
                trace = true;
            } else if (*nextArgP == "-n" || *nextArgP == "--count") {
                nextArgP = args.next();
                if (!from_string<int> (maxCount, *nextArgP) || maxCount <= 0) {
                    throw "Count must be a positive integer and not: " + *nextArgP;
                }
            } else if (*nextArgP == "--url") {
                nextArgP = args.next();
                serverUrl = *nextArgP;
            } else {
                throw "Unexpected switch '" + *nextArgP + "'";
            }
            nextArgP = args.next();
        }

        if (!nextArgP) {
            throw usage;
        }
        std::string model = *nextArgP;
        nextArgP = args.next();

        std::string entityType;
        std::string filter;
        saga::session sess;
        std::string attrs;

        if (!nextArgP) {
            throw usage;
        }
        entityType = *nextArgP;
        nextArgP = args.next();
        filter = ""; /* set the default value */
        if (nextArgP && (*nextArgP)[0] == '[') {
            filter = nextArgP->substr(1, nextArgP->size() - 2);
            nextArgP = args.next();
        }
        if (nextArgP && (*nextArgP)[0] == '(') {
            attrs = nextArgP->substr(1, nextArgP->size() - 2);
            nextArgP = args.next();
            if (nextArgP) {
                throw "Attribute list may only be at the end";
            }
        }

        saga::isn::entity_data_set eds(model, entityType, filter, sess, serverUrl);
        if (trace) {
            printTrace(eds, entityType, filter);
        }

        while (nextArgP && !eds.get_data().empty()) {
            entityType = *nextArgP;
            nextArgP = args.next();
            filter = ""; /* set the default value */
            if (nextArgP && (*nextArgP)[0] == '[') {
                filter = nextArgP->substr(1, nextArgP->size() - 2);
                nextArgP = args.next();
            }
            if (nextArgP && (*nextArgP)[0] == '(') {
                attrs = nextArgP->substr(1, nextArgP->size() - 2);
                nextArgP = args.next();
                if (nextArgP) {
                    throw "Attribute list may only be at the end";
                }
            }
            eds = eds.get_related_entities(entityType, filter);

            if (trace) {
                printTrace(eds, entityType, filter);
            }
        }
        std::vector<std::string> attListRequested;
        tokenize(attrs, attListRequested, ", ");
        std::vector<saga::isn::entity_data> data = eds.get_data();
        std::vector<std::string> attList;
        for (int i = 0; i < data.size() && maxCount > 0; i++, --maxCount) {
            if (attListRequested.empty()) {
                attList = data[i].list_attributes();
            } else {
                attList = attListRequested;
            }
            bool spaceNeeded = attListRequested.size() != 1;
            for (int j = 0; j < attList.size(); j++) {
                try {
                    if (!data[i].attribute_is_vector(attList[j])) {
                        std::string attribValue = data[i].get_attribute(attList[j]);
                        if (spaceNeeded) {
                            std::cout << std::endl;
                            spaceNeeded = false;
                        }
                        std::cout << attList[j] << " : " << attribValue << std::endl;
                    } else {
                        std::vector<std::string> attribValues = data[i].get_vector_attribute(attList[j]);
                        if (spaceNeeded) {
                            std::cout << std::endl;
                            spaceNeeded = false;
                        }
                        for (int k = 0; k < attribValues.size(); k++) {
                            std::cout << attList[j] << " : " << attribValues[k] << std::endl;
                        }
                    }
                } catch (saga::does_not_exist& e) {
                    /* ignore it */
                }
            }
        }
    } catch (const std::string& str) {
        std::cerr << str << std::endl;
        return 1;
    } catch (char const * str) {
        std::cerr << str << std::endl;
        return 1;
    } catch (saga::exception& e) {
        std::cerr << e.what() << std::endl;
        return 1;
    }

    return 0;
#endif
}
